From c41adeec0299569b307424b75a3c20665e6c8172 Mon Sep 17 00:00:00 2001
From: Vitaly Kuznetsov <vkuznets@redhat.com>
Date: Thu, 10 Jul 2014 15:24:08 -0400
Subject: [virt] xen-blkfront: handle backend CLOSED without CLOSING

Message-id: <1405005849-29739-2-git-send-email-vkuznets@redhat.com>
Patchwork-id: 85665
O-Subject: [RHEL7.1 PATCH 1/2] xen-blkfront: handle backend CLOSED without CLOSING
Bugzilla: 1070868
RH-Acked-by: Andrew Jones <drjones@redhat.com>
RH-Acked-by: Igor Mammedov <imammedo@redhat.com>
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1070868
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7684646
Tested: locally

Upstream commit: 3661371701e714f0cea4120f6a365340858fb4e4
Author: David Vrabel <david.vrabel@citrix.com>
Date:   Tue Feb 4 18:53:56 2014 +0000

Backend drivers shouldn't transistion to CLOSED unless the frontend is
CLOSED.  If a backend does transition to CLOSED too soon then the
frontend may not see the CLOSING state and will not properly shutdown.

So, treat an unexpected backend CLOSED state the same as CLOSING.

Signed-off-by: David Vrabel <david.vrabel@citrix.com>
Acked-by: Konrad Rzeszutek Wilk <konrad.wilk@oracle.com>
Cc: stable@vger.kernel.org
Signed-off-by: Konrad Rzeszutek Wilk <konrad.wilk@oracle.com>
Signed-off-by: Vitaly Kuznetsov <vkuznets@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/block/xen-blkfront.c b/drivers/block/xen-blkfront.c
index a4660bb..c89618b 100644
--- a/drivers/block/xen-blkfront.c
+++ b/drivers/block/xen-blkfront.c
@@ -1852,13 +1852,16 @@ static void blkback_changed(struct xenbus_device *dev,
  case XenbusStateReconfiguring:
  case XenbusStateReconfigured:
  case XenbusStateUnknown:
- case XenbusStateClosed:
   break;
 
  case XenbusStateConnected:
   blkfront_connect(info);
   break;
 
+ case XenbusStateClosed:
+  if (dev->state == XenbusStateClosed)
+   break;
+  /* Missed the backend's Closing state -- fallthrough */
  case XenbusStateClosing:
   blkfront_closing(info);
   break;
-- 
1.7.1